gapis/api/transform2/capture_log.go

```diff
@@ -0,0 +1,130 @@
+// Copyright (C) 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transform2
+
+import (
+	"context"
+	"os"
+
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/core/memory/arena"
+	"github.com/google/gapid/gapis/api"
+	"github.com/google/gapid/gapis/capture"
+)
+
+type captureLog struct {
+	file   *os.File
+	header *capture.Header
+	cmds   []api.Cmd
+	ids    map[api.CmdID]api.CmdID
+}
+
+// NewCaptureLog returns a Transformer that will log all commands passed through it
+// to the capture file at path.
+func NewCaptureLog(ctx context.Context, sourceCapture *capture.GraphicsCapture, path string) *captureLog {
+	f, err := os.Create(path)
+	if err != nil {
+		log.E(ctx, "Failed to create replay capture file %v: %v", path, err)
+		return nil
+	}
+	return &captureLog{
+		file:   f,
+		header: sourceCapture.Header,
+		cmds:   []api.Cmd{},
+		ids:    map[api.CmdID]api.CmdID{},
+	}
+}
+
+func (logTransform *captureLog) RequiresAccurateState() bool {
+	return false
+}
+
+func (logTransform *captureLog) BeginTransform(ctx context.Context, inputCommands []api.Cmd, inputState *api.GlobalState) ([]api.Cmd, error) {
+	return inputCommands, nil
+}
+
+func (logTransform *captureLog) ClearTransformResources(ctx context.Context) {
+	// Do nothing
+}
+
+func (logTransform *captureLog) TransformCommand(ctx context.Context, id api.CmdID, inputCommands []api.Cmd, inputState *api.GlobalState) ([]api.Cmd, error) {
+	// Melih TODO: We cannot have this check anymore.
```

Comment:
When we have two transform chains, the transform itself will only see real or initial commands as long as we do not reuse transforms between chains. That means we can store a flag inside the transform when it is created saying that behaviour it should take, and it can just do that without worrying about command IDs

Reply:
This actually have another problem that we have forgotten to mention. This transform has to run for both initial commands and real commands without breaking(as it logs all the commands and creates a new trace from them). This is a thing that we cannot do right now.

One solution can be depending on the flag, we can prefer to do nothing on Begin or End Transform but I have serious drawbacks about allowing that behaviour as it makes it effectively a transform with two pieces.

I am leaving this for now. We can talk about it when I am back from vacation.